src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs

```diff
@@ -0,0 +1,350 @@
+            int decIndex = s.IndexOf('.');
+            if (decIndex <= 0)
+                return 0;
+            return len - decIndex - 1;
+        }
+
+        /// <summary>
+        /// Applies the inherent scale calculation 
+        /// to every ordinate in a geometry.
+        /// <para/>
+        /// WARNING: this is <b>very</b> slow.
+        /// </summary>
+        /// <author>Martin Davis</author>
+        private class InherentScaleFilter : IEntireCoordinateSequenceFilter
+        {
+
+            private double _scale = 0;
+
+            public double Scale
+            {
+                get => _scale;
+            }
+
+            public void Filter(CoordinateSequence sequence)
+            {
+                for (int i = 0; i < sequence.Count; i++)
+                {
+                    UpdateScaleMax(sequence.GetX(i));
+                    UpdateScaleMax(sequence.GetY(i));
+                }
+            }
+
+            private void UpdateScaleMax(double value)
+            {
+                double scaleVal = PrecisionUtility.InherentScale(value);
+                if (scaleVal > _scale)
+                {
+                    //System.out.println("Value " + value + " has scale: " + scaleVal);
+                    _scale = scaleVal;
+                }
+            }
+
+            public bool Done
+            {
+                get => false;
+            }
+
+            public bool GeometryChanged
+            {
+                get => false;
+            }
+        }
+
+        /// <summary>
+        /// Determines a precision model to
+        /// use for robust overlay operations for one geometry.
+        /// The precision scale factor is chosen to maximize
+        /// output precision while avoiding round-off issues.
+        /// <para/>
+        /// NOTE: this is a heuristic determination, so is not guaranteed to
+        /// eliminate precision issues.
+        /// <para/>
+        /// WARNING: this is <b>very</b> slow.
+        /// </summary>
+        /// <param name="a">A geometry</param>
+        /// <returns>A suitable precision model for overlay</returns>
+        public static PrecisionModel RobustPM(Geometry a)
+        {
+            double scale = PrecisionUtility.RobustScale(a);
+            return new PrecisionModel(scale);
+        }
+
+        /// <summary>
+        /// Determines a scale factor which maximizes
+        /// the digits of precision and is
+        /// safe to use for overlay operations.
+        /// The robust scale is the minimum of the
+        /// inherent scale and the safe scale factors.
+        /// <para/>
+        /// WARNING: this is <b>very</b> slow.
+        /// </summary>
+        /// <param name="a">A geometry</param>
+        /// <param name="b">A geometry</param>
+        /// <returns>A scale factor for use in overlay operations</returns>
+        public static double RobustScale(Geometry a, Geometry b)
+        {
+            double inherentScale = InherentScale(a, b);
```

Comment:
validate `a` non-null before using it

(we could rely on `InherentScale` to validate for us if we do an earlier comment, mainly just mentioning this here for consistency)